parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.kt

```diff
@@ -0,0 +1,89 @@
+package com.parsely.parselyandroid
+
+import kotlin.time.Duration
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+/**
+ * Engagement manager for article and video engagement.
+ *
+ *
+ * Implemented to handle its own queuing of future executions to accomplish
+ * two things:
+ *
+ *
+ * 1. Flushing any engaged time before canceling.
+ * 2. Progressive backoff for long engagements to save data.
+ */
+internal class EngagementManager(
+    private val parselyTracker: ParselyTracker,
+    private var latestDelayMillis: Long,
+    private val baseEvent: Map<String, Any>,
+    private val intervalCalculator: HeartbeatIntervalCalculator,
+    private val coroutineScope: CoroutineScope,
+    private val clock: Clock,
+) {
+    private var job: Job? = null
+    private var totalTime: Long = 0
+    private var startTime: Duration
+    private var nextScheduledExecution: Long = 0
+
+    init {
+        startTime =  clock.now
+    }
+
+    val isRunning: Boolean
+        get() = job?.isActive ?: false
+
+    fun start() {
+        startTime = clock.now
+        job = coroutineScope.launch {
+            while (isActive) {
+                latestDelayMillis = intervalCalculator.calculate(startTime)
+                nextScheduledExecution = clock.now.inWholeMilliseconds + latestDelayMillis
+                delay(latestDelayMillis)
+                doEnqueue(clock.now.inWholeMilliseconds)
+            }
+        }
+    }
+
+    fun stop() {
+        job?.let {
+            it.cancel()
+            doEnqueue(nextScheduledExecution)
```

Comment:
Could you provide me some context about what this line is accomplishing, so I can provide better feedback?

Reply:
Thanks a lot for the context!

> There's also a possibility, that the original implementation was enqueuing this event unintended - do you think it's what happened here?

I thought that might be possible - but the previous implementation looks pretty intentional.

---

This logic feels a little fragile and out of place, but as we discussed previously, it makes sense for this PR to closely match the previous logic. So, I don't think we need to do anything about that at the moment.

Adding a comment about what's happening _could_ be useful, but I'll admit that I still don't full get it even with your explanation. I'd need to dive deeper into what each component is doing and how it's utilized, but that's not really worth it at the moment.

---

@wzieba I don't have any other comments that needs addressing, so please feel free to proceed with merging the PR.